File: src/leetCodeProblems/BinaryTree/BinaryTreeLCA236.java
package leetCodeProblems.BinaryTree;

public class BinaryTreeLCA236 {
	static boolean isN1Found = false;
    static boolean isN2Found = false;
    
    static class TreeNode {
    	int val;
        TreeNode left;
        TreeNode right;

        TreeNode(int val) {
            this.val = val;
        }

        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }
    
    // node is nothing BUT root of that recursive TREE.
    TreeNode findLCAUtil(TreeNode node, int n1, int n2) {

        if (node == null) {
            return null;
        }

        TreeNode temp = null;
        
        // Root matches with n1
        if (node.val == n1) {
            temp = node;
            isN1Found = true;
        }
        
        // Root matches with n2
        if (node.val == n2) {
            temp = node;
            isN2Found = true;
        }

        if (temp != null) {
            return temp;
        }
        
        // Either n1 or n2 exits as CHILDs in the left child of node
        TreeNode leftLCA = findLCAUtil(node.left, n1, n2); 
        TreeNode rightLCA = findLCAUtil(node.right, n1, n2);

        // LCA is found - This is IMPORTANT code
        if (leftLCA != null && rightLCA != null) {
            return node; // Note we are returning parent of leftLCA & rightLCA. NOT leftLCA or rightLCA.
        }

        if (leftLCA != null) {
            return leftLCA; // Return leftLCA, not PARENT.
        }
        else {
            return rightLCA;
        }
    }

    public int lca(TreeNode A, int B, int C) {
    	isN1Found = isN2Found = false;

        TreeNode LCA = findLCAUtil(A, B, C);

        if (isN1Found && isN2Found) {
            return LCA.val;
        }
        else {
            return -1;
        }
    }
    
    public static void main(String[] args) {
        
    	TreeNode root = new TreeNode(1);
        root.left = new TreeNode(2);
        root.right = new TreeNode(3);
        root.left.left = new TreeNode(4);
        root.left.right = new TreeNode(5);
        root.right.left = new TreeNode(6);
        root.right.right = new TreeNode(7);
        
        BinaryTreeLCA236 ob = new BinaryTreeLCA236();

        int output = ob.lca(root, 4, 7);
        
        System.out.println(output);

    }

}

File: src/leetCodeProblems/BinaryTree/BinaryTreeLCA236.go
package main

/*
- Leetcode - https://leetcode.com/problems/lowest-common-ancestor-of-a-binary-tree/description/
- Time - O(n)
- Space - O(1)
*/

func lowestCommonAncestor(root, p, q *TreeNode) *TreeNode {
	if root == nil {
		return nil
	}

	if root == p {
		return root
	}

	if root == q {
		return root
	}

	leftLCA := lowestCommonAncestor(root.Left, p, q)
	rightLCA := lowestCommonAncestor(root.Right, p, q)

	if leftLCA != nil && rightLCA != nil {
		return root
	}

	if leftLCA != nil {
		return leftLCA
	} else {
		return rightLCA
	}
}

// func main() {
// 	root := new(TreeNode) //Returns pointer to TreeNode object
// 	root.Val = 3
// 	root.Left = new(TreeNode)
// 	root.Left.Val = 5
// 	root.Left.Left = new(TreeNode)
// 	root.Left.Left.Val = 6
// 	root.Left.Right = new(TreeNode)
// 	root.Left.Right.Val = 2
// 	root.Left.Right.Left = new(TreeNode)
// 	root.Left.Right.Left.Val = 7
// 	root.Left.Right.Right = new(TreeNode)
// 	root.Left.Right.Right.Val = 4

// 	root.Right = new(TreeNode)
// 	root.Right.Val = 1
// 	root.Right.Left = new(TreeNode)
// 	root.Right.Left.Val = 0
// 	root.Right.Right = new(TreeNode)
// 	root.Right.Right.Val = 8

// 	//log.Println(lowestCommonAncestor(root, root.Left, root.Right))
// 	log.Println(lowestCommonAncestor(root, root.Left, root.Left.Right.Right))

// }
